File: _code-samples/escrow/py/cancel_escrow.py
import json
from datetime import datetime, timedelta, UTC
from time import sleep

from xrpl.clients import JsonRpcClient
from xrpl.models import EscrowCreate, EscrowCancel
from xrpl.models.requests import AccountObjects, Ledger, Tx
from xrpl.transaction import submit_and_wait
from xrpl.utils import datetime_to_ripple_time, ripple_time_to_datetime, get_balance_changes
from xrpl.wallet import generate_faucet_wallet

# Set up client and get a wallet
client = JsonRpcClient("https://s.altnet.rippletest.net:51234")
print("Funding new wallet from faucet...")
wallet = generate_faucet_wallet(client, debug=True)
# destination_address = "rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe" # Testnet faucet
# Alternative: Get another account to send the escrow to. Use this if you get
# a tecDIR_FULL error trying to create escrows to the Testnet faucet.
destination_address = generate_faucet_wallet(client, debug=True).address

# Create an escrow that won't be finished --------------------------------------
cancel_delay = 30
cancel_after = datetime.now(tz=UTC) + timedelta(seconds=cancel_delay)
print("This escrow will expire after", cancel_after)
cancel_after_rippletime = datetime_to_ripple_time(cancel_after)
# Use a crypto-condition that nobody knows the fulfillment for
condition_hex = "A02580200000000000000000000000000000000000000000000000000000000000000000810120"
escrow_create = EscrowCreate(
    account=wallet.address,
    destination=destination_address,
    amount="123456", # drops of XRP
    condition=condition_hex,
    cancel_after=cancel_after_rippletime
)
print("Signing and submitting the EscrowCreate transaction.")
response = submit_and_wait(escrow_create, client, wallet, autofill=True)
print(json.dumps(response.result, indent=2))

result_code = response.result["meta"]["TransactionResult"]
if result_code != "tesSUCCESS":
    print(f"EscrowCreate failed with result code {result_code}")
    exit(1)

# Wait for the escrow to expire ------------------------------------------------
# Since ledger close times can be rounded by up to 10 seconds, wait an extra
# 10 seconds to make sure the escrow has officially expired.
print(f"Waiting {cancel_delay + 10} seconds for the escrow to expire.")
sleep(cancel_delay + 10)

# Look up the official close time of the validated ledger ----------------------
validated_ledger = client.request(Ledger(ledger_index="validated"))
close_time = validated_ledger.result["ledger"]["close_time"]
print("Latest validated ledger closed at", 
    ripple_time_to_datetime(close_time)
)
ledger_hash = validated_ledger.result["ledger"]["ledger_hash"]

# Look up escrows connected to the account, handling pagination ----------------
expired_escrow = None
marker = None
while True:
    try:
        response = client.request(AccountObjects(
            account=wallet.address,
            ledger_hash=ledger_hash,
            type="escrow",
            marker=marker
        ))
    except Exception as e:
        print(f"Error: account_objects failed: {e}")
        exit(1)

    for escrow in response.result["account_objects"]:
        if "CancelAfter" not in escrow:
            print("This escrow does not have an expiration")
        elif escrow["CancelAfter"] < close_time:
            print("This escrow has expired.")
            expired_escrow = escrow
            break
        else:
            expiration_time = ripple_time_to_datetime(escrow["CancelAfter"])
            print(f"This escrow expires at {expiration_time}.")
    
    if expired_escrow:
        # Found an expired escrow, stop paginating
        break
    
    if "marker" in response.result.keys():
        marker=marker
    else:
        # This is the last page of results
        break

if not expired_escrow:
    print("Did not find any expired escrows.")
    exit(1)

# Find the sequence number of the expired escrow -------------------------------
response = client.request(Tx(transaction=expired_escrow["PreviousTxnID"]))
if not response.is_successful():
    print("Couldn't get transaction. Maybe this server doesn't have enough "
            "transaction history available?")
    exit(1)

if response.result["tx_json"]["TransactionType"] == "EscrowCreate":
    # Save this sequence number for canceling the escrow
    escrow_seq = response.result["tx_json"]["Sequence"]
    if escrow_seq == 0:
        # This transaction used a Ticket, so use the TicketSequence instead.
        escrow_seq = response.result["tx_json"]["TicketSequence"]
else:
    # Currently, this is impossible since no current transaction can update
    # an escrow without finishing or canceling it. But in the future, if
    # that becomes possible, you would have to look at the transaction
    # metadata to find the previous transaction and repeat until you found
    # the transaction that created the escrow.
    print("The escrow's previous transaction wasn't EscrowCreate!")
    exit(1)

# Send EscrowCancel transaction ------------------------------------------------
escrow_cancel = EscrowCancel(
    account=wallet.address,
    owner=expired_escrow["Account"],
    offer_sequence=escrow_seq
)
print("Signing and submitting the EscrowCancel transaction.")
response2 = submit_and_wait(escrow_cancel, client, wallet, autofill=True)
print(json.dumps(response2.result, indent=2))

result_code = response2.result["meta"]["TransactionResult"]
if result_code != "tesSUCCESS":
    print(f"EscrowCancel failed with result code {result_code}")
    exit(1)

print("Escrow canceled. Balance changes:")
print(json.dumps(get_balance_changes(response2.result["meta"]), indent=2))

File: _code-samples/escrow/js/cancel-escrow.js
import { Client, isoTimeToRippleTime, rippleTimeToISOTime, validate, getBalanceChanges } from 'xrpl'

const client = new Client('wss://s.altnet.rippletest.net:51233')
await client.connect()

console.log('Funding new wallet from faucet...')
const { wallet } = await client.fundWallet()
// const destinationAddress = 'rPT1Sjq2YGrBMTttX4GZHjKu9dyfzbpAYe' // Testnet faucet
// Alternative: Get another account to send the escrow to. Use this if you get
// a tecDIR_FULL error trying to create escrows to the Testnet faucet.
const destinationAddress = (await client.fundWallet()).wallet.address

// Create an escrow that won't be finished -------------------------------------
const cancelDelay = 30
const cancelAfter = new Date()
cancelAfter.setSeconds(cancelAfter.getSeconds() + cancelDelay)
console.log('This escrow will expire after:', cancelAfter)
// Convert cancelAfter to seconds since the Ripple Epoch:
const cancelAfterRippleTime = isoTimeToRippleTime(cancelAfter.toISOString())
const conditionHex = 'A02580200000000000000000000000000000000000000000000000000000000000000000810120'

const escrowCreate = {
  TransactionType: 'EscrowCreate',
  Account: wallet.address,
  Destination: destinationAddress,
  Amount: '123456',
  Condition: conditionHex,
  CancelAfter: cancelAfterRippleTime
}
validate(escrowCreate)

console.log('Signing and submitting the EscrowCreate transaction.')
const response = await client.submitAndWait(escrowCreate, {
  wallet,
  autofill: true // Note: fee is higher based on condition size in bytes
})
console.log(JSON.stringify(response.result, null, 2))
const escrowCreateResultCode = response.result.meta.TransactionResult
if (escrowCreateResultCode !== 'tesSUCCESS') {
  console.error(`EscrowCreate failed with code ${escrowCreateResultCode}.`)
  client.disconnect()
  process.exit(1)
}

// Wait for the escrow to expire -----------------------------------------------
// Since ledger close times can be rounded by up to 10 seconds, wait an extra
// 10 seconds to make sure the escrow has officially expired.
console.log(`Waiting ${cancelDelay + 10} seconds for the escrow to expire...`)
await sleep(cancelDelay + 10)

/* Sleep function that can be used with await */
function sleep (delayInSeconds) {
  const delayInMs = delayInSeconds * 1000
  return new Promise((resolve) => setTimeout(resolve, delayInMs))
}

// Look up the official close time of the validated ledger ---------------------
const ledger = await client.request({
  command: 'ledger',
  ledger_index: 'validated'
})
if (ledger.error) {
  console.error(`Error looking up validated ledger: ${ledger.error}`)
  client.disconnect()
  process.exit(1)
}
const closeTime = ledger.result.ledger.close_time
console.log('Latest validated ledger closed at',
  rippleTimeToISOTime(closeTime)
)
const ledgerHash = ledger.result.ledger.ledger_hash

// Look up escrows connected to the account, handling pagination ---------------
let marker
let expiredEscrow
while (true) {
  console.log(`Requesting page of account_objects with marker ${marker}`)
  const resp = await client.request({
    command: 'account_objects',
    account: wallet.address,
    ledger_hash: ledgerHash,
    type: 'escrow',
    marker
  })
  if (resp.error) {
    console.error('account_objects failed with error', resp)
    client.disconnect()
    process.exit(1)
  }

  // Add new escrows to the full list
  for (const escrow of resp.result.account_objects) {
    if (!escrow.hasOwnProperty('CancelAfter')) {
      console.log('This escrow does not have an expiration.')
    } else if (escrow.CancelAfter < closeTime) {
      console.log('This escrow has expired.')
      expiredEscrow = escrow
      break
    } else {
      const expirationTime = rippleTimeToISOTime(escrow.CancelAfter)
      console.log('This escrow expires at', expirationTime)
    }
  }

  if (expiredEscrow) {
    // Found an expired escrow, stop paginating
    break
  }

  // If there's a marker, loop and fetch the next page of results
  if (resp.result.marker) {
    marker = resp.result.marker
  } else {
    break
  }
}

if (!expiredEscrow) {
  console.error('Did not find any expired escrows.')
  process.exit(1)
}

// Find the sequence number of the expired escrow ------------------------------
let escrow_seq
const txResp = await client.request({
  command: 'tx',
  transaction: expiredEscrow.PreviousTxnID
})
if (txResp.error) {
  console.error("Couldn't get transaction. Maybe this server doesn't have",
    'enough transaction history available?')
  client.disconnect()
  process.exit(1)
}

if (txResp.result.tx_json.TransactionType === 'EscrowCreate') {
  // Save this sequence number for canceling the escrow
  escrow_seq = txResp.result.tx_json.Sequence
  if (escrow_seq === 0) {
    // This transaction used a Ticket, so use TicketSequence instead.
    escrow_seq = response.result.tx_json.TicketSequence
  }
} else {
  console.error("This escrow's previous transaction wasn't EscrowCreate!")
  client.disconnect()
  process.exit(1)
}

// Send EscrowCancel transaction -----------------------------------------------
const escrowCancel = {
  TransactionType: 'EscrowCancel',
  Account: wallet.address,
  Owner: expiredEscrow.Account,
  OfferSequence: escrow_seq
}
validate(escrowCancel)

console.log('Signing and submitting the EscrowCancel transaction.')
const cancelResponse = await client.submitAndWait(escrowCancel, {
  wallet,
  autofill: true
})
console.log(JSON.stringify(cancelResponse.result, null, 2))
const cancelResultCode = cancelResponse.result.meta.TransactionResult
if (cancelResultCode !== 'tesSUCCESS') {
  console.error(`EscrowCancel failed with result code ${cancelResultCode}`)
  client.disconnect()
  process.exit(1)
}

console.log('Escrow canceled. Balance changes:')
console.log(JSON.stringify(getBalanceChanges(cancelResponse.result.meta), null, 2))

client.disconnect()
